core/src/main/java/de/jplag/merging/MatchMerging.java

```diff
@@ -0,0 +1,136 @@
+        long timeBeforeStartInMillis = System.currentTimeMillis();
+        for (int i = 0; i < comparisons.size(); i++) {
+            firstSubmission = comparisons.get(i).firstSubmission().copy();
+            secondSubmission = comparisons.get(i).secondSubmission().copy();
+            globalMatches = new ArrayList<>(comparisons.get(i).matches());
+            globalMatches.addAll(comparisons.get(i).ignoredMatches());
+            computeNeighbors();
+            mergeNeighbors();
+            removeBuffer();
+            comparisons.set(i, new JPlagComparison(firstSubmission, secondSubmission, globalMatches, new ArrayList<>()));
+
+        }
+        long durationInMillis = System.currentTimeMillis() - timeBeforeStartInMillis;
+        return new JPlagResult(comparisons, result.getSubmissions(), result.getDuration() + durationInMillis, options);
+    }
+
+    private void computeNeighbors() {
+        neighbors = new ArrayList<>();
+        List<Match> sortedByFirst = new ArrayList<>(globalMatches);
+        Collections.sort(sortedByFirst, (m1, m2) -> m1.startOfFirst() - m2.startOfFirst());
+        List<Match> sortedBySecond = new ArrayList<>(globalMatches);
+        Collections.sort(sortedBySecond, (m1, m2) -> m1.startOfSecond() - m2.startOfSecond());
+        for (int i = 0; i < sortedByFirst.size() - 1; i++) {
+            if (sortedBySecond.indexOf(sortedByFirst.get(i)) == (sortedBySecond.indexOf(sortedByFirst.get(i + 1)) - 1)) {
+                neighbors.add(Arrays.asList(sortedByFirst.get(i), sortedByFirst.get(i + 1)));
+            }
+        }
+    }
+
+    private void mergeNeighbors() {
+        int i = 0;
+        while (i < neighbors.size()) {
+            int lengthUpper = neighbors.get(i).get(0).length();
+            int lengthLower = neighbors.get(i).get(1).length();
+            int seperatingFirst = neighbors.get(i).get(1).startOfFirst() - neighbors.get(i).get(0).endOfFirst() - 1;
+            int seperatingSecond = neighbors.get(i).get(1).startOfSecond() - neighbors.get(i).get(0).endOfSecond() - 1;
+            double seperating = (seperatingFirst + seperatingSecond) / 2.0;
+            // Checking length is not necessary as GST already checked length while computing matches
+            if (seperating <= seperatingThreshold) {
+                globalMatches.removeAll(neighbors.get(i));
+                globalMatches
+                        .add(new Match(neighbors.get(i).get(0).startOfFirst(), neighbors.get(i).get(0).startOfSecond(), lengthUpper + lengthLower));
+                removeToken(neighbors.get(i).get(0).startOfFirst(), neighbors.get(i).get(0).startOfSecond(), lengthUpper, seperatingFirst,
+                        seperatingSecond);
+                computeNeighbors();
+                i = 0;
+            } else {
+                i++;
+            }
+        }
+    }
+
+    private void removeToken(int startFirst, int startSecond, int lengthUpper, int seperatingFirst, int seperatingSecond) {
```

Comment:
Also here JDoc.

Reply:
For these three parameters it is hard to grasp what they are due to their name: int lengthUpper, int seperatingFirst, int seperatingSecond.